### PR TITLE
Add content type for easy-to-consume JSON

### PR DESCRIPTION
**Pull request by jameshaydon**
Adds a `application/radicle-json` content-type so that consumers of the `query` endpoint may accept conveniently encoded JSON. E.g. for the `radicle-garden`:

`(list-patches)`:

```json
{
  "15": {
    "author-email": "haze",
    "state": "accepted",
    "created-at": "2019-03-13T16:00:51Z",
    "patch": "From 7c1002a7c13666a377b98241e87e699178ba0d67 Mon Sep 17 00:00:00 2001\nFrom: haze <isnt@haze.cool>\nDate: Wed, 13 Mar 2019 12:00:41 -0400\nSubject: [PATCH] Uprooting a galaxy\n\n---\n 2019-03-13_haze.txt | 8 ++++++++\n 1 file changed, 8 insertions(+)\n create mode 100644 2019-03-13_haze.txt\n\ndiff --git a/2019-03-13_haze.txt b/2019-03-13_haze.txt\nnew file mode 100644\nindex 0000000..80e8c14\n--- /dev/null\n+++ b/2019-03-13_haze.txt\n@@ -0,0 +1,8 @@\n+┌──────────────────┐\n+│       \\Λ/        │\n+│     ──▕ ▏──      │\n+│       /V\\        │\n+│                  │\n+│  T A C H Y O N   │\n+│ S O F T W A R E  │\n+└──────────────────┘\n\\ No newline at end of file\n-- \n2.17.2 (Apple Git-113)\n",
    "author-name": "haze",
    "author": [
      "public-key",
      {
        "public_curve": [
          "curve-fP",
          [
            "curve-prime",
            115792089237316195423570985008687907853269984665640564039457584007908834671663,
            [
              "curve-common",
              {
                "ecc_g": [
                  "point",
                  55066263022277343669578718895168534326250603453777594175500187360389116729240,
                  32670510020758816978083085130507043184471273380659243275938904335757337482424
                ],
                "ecc_a": 0,
                "ecc_n": 115792089237316195423570985008687907852837564279074904382605163141518161494337,
                "ecc_h": 1,
                "ecc_b": 7
              }
            ]
          ]
        ],
        "public_q": [
          "point",
          84104661149731357559625146477383645349462680667189717717948604614340334257896,
          28879832228416370287593549504197030105718169503737631177310776616050059709102
        ]
      }
    ],
    "short-hash": "7c1002a",
    "comments": [],
    "title": "Uprooting a galaxy",
    "number": 15,
    "description": "",
    "modified-at": "2019-03-13T16:29:55Z"
  },
  "more": "..."
}
```

### CODE REVIEW

File: src/Radicle/Daemon/HttpApi.hs [hunk 1]
```diff
@@ -39,51 +77,71 @@ instance FromHttpApiData MachineId where
...
+
+type QueryRequest = QueryRequestF Value
 
-instance A.ToJSON QueryRequest where
+instance {-# OVERLAPPING #-} A.ToJSON QueryRequest where
     toJSON QueryRequest{..} = A.object
         [ "expression" .= valueToJson expression
         ]
-instance A.FromJSON QueryRequest where
+instance {-# OVERLAPPING #-} A.FromJSON QueryRequest where
```
Comment (jkarni) [inline src/Radicle/Daemon/HttpApi.hs]: What is this overlapping with?

Comment (jameshaydon) [reply src/Radicle/Daemon/HttpApi.hs]: https://github.com/radicle-dev/radicle/pull/586/files#diff-7a716c0d456d68b29a5cdf5fa62a6bf0R81

Which I add to be able to encode all the Req/Resp types in one go:
https://github.com/radicle-dev/radicle/pull/586/files#diff-7a716c0d456d68b29a5cdf5fa62a6bf0R70

I also tried to then encode/decode the JSON for the Req/Resp types in one go but got stuck with an overlap for `[a]`.

File: src/Radicle/Internal/Json.hs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import           Radicle.Internal.Identifier
+import           Radicle.Internal.Number
+import           Radicle.Internal.Parse
+import           Radicle.Internal.Pretty
+import qualified Radicle.Internal.Type as Type
+
+-- | Parses JSON into a Value. It expects the value to be encoded as a radicle
+-- expression in a JSON string. Corresponds to the encoding given by
+-- 'valueToJson'.
+jsonToValue :: A.Value -> A.Parser Value
```
Comment (jkarni) [inline src/Radicle/Internal/Json.hs]: Can you give a doctest example?

File: src/Radicle/Internal/Json.hs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+jsonToValue = A.withText "Value" $ \t -> do
+    case parse "[daemon]" t of
+      Left err -> fail $ "failed to parse Radicle expression: " <> show err
+      Right v  -> pure v
+
+-- | Encodes a radicle value as JSON. The value is encoded a radicle expression
```
Comment (jkarni) [inline src/Radicle/Internal/Json.hs]: Don't know how to do multiline suggestions, but maybe instead:
```
-- | Pretty-prints a Radicle value, and wraps it in a JSON string
```
So it's clear this isn't in "deeper" way encoding Rad as JSON.

File: src/Radicle/Internal/Json.hs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    jsonKey :: Value -> Either Text Text
+    jsonKey (String s)          = pure s
+    jsonKey (Keyword (Ident i)) = pure i
+    jsonKey (Atom (Ident i))    = pure i
+    jsonKey n@(Number _)        = pure (renderCompactPretty n)
+    jsonKey (Boolean b)         = pure $ if b then "true" else "false"
+    jsonKey v                   = Left $ cantConvertType v "a JSON key"
+
+    cantConvertType :: Value -> Text -> Text
+    cantConvertType v toThis = "Can not convert values of type " <> toS (Type.typeString (valType v)) <> " to " <> toThis
```
Comment (jkarni) [inline src/Radicle/Internal/Json.hs]:
```suggestion
    cantConvertType v toThis = "Cannot convert values of type " <> toS (Type.typeString (valType v)) <> " to " <> toThis
```

File: src/Radicle/Internal/Json.hs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+
+-- | Converts an `aeson` value to a radicle one.
+fromJson :: A.Value -> Value
+fromJson = \case
+  A.Number n -> Number (toRational n) -- TODO(james): think about untrusted sources.
```
Comment (jkarni) [inline src/Radicle/Internal/Json.hs]: Should we create an issue instead? TODO's usually get lost....

Comment (jameshaydon) [reply src/Radicle/Internal/Json.hs]: No I will just make it safer now.

File: src/Radicle/Internal/Json.hs (hunk 3 above)
Comment (jkarni) [reply src/Radicle/Internal/Json.hs]: And maybe also a doctest example

File: src/Radicle/Daemon/HttpApi.hs [hunk 6]
```diff
@@ -25,12 +32,43 @@ import           Data.Aeson ((.:), (.=))
...
+instance Accept RadicleData where
+  contentType _ = "application" HttpMedia.// "radicle"
+
+instance ToRad Ann.WithPos a => MimeRender RadicleData a where
+  mimeRender _ = toS . renderCompactPretty . (toRad :: a -> Value)
+
+instance FromRad Ann.WithPos a => MimeUnrender RadicleData a where
+  mimeUnrender _ = first toS . (fromRad <=< parse "[daemon]") . toS
+
+-- | Content type for values encoded *loosely* as JSON.
```
Comment (jkarni) [inline src/Radicle/Daemon/HttpApi.hs]: Maybe a bit more explanation as to what this means?

File: src/Radicle/Daemon/HttpApi.hs [hunk 7]
```diff
@@ -39,51 +77,71 @@ instance FromHttpApiData MachineId where
...
-  deriving (Generic)
+newtype QueryRequestF a = QueryRequest { expression :: a }
+  deriving (Functor, Foldable, Traversable, Generic, A.ToJSON, A.FromJSON)
+
+type QueryRequest = QueryRequestF Value
 
-instance A.ToJSON QueryRequest where
+instance {-# OVERLAPPING #-} A.ToJSON QueryRequest where
     toJSON QueryRequest{..} = A.object
         [ "expression" .= valueToJson expression
```
Comment (jkarni) [inline src/Radicle/Daemon/HttpApi.hs]: It seems a little ugly to have two different ways of encoding to JSON, and then picking one arbitrarily as the instance...

File: src/Radicle/Daemon/HttpApi.hs [hunk 8]
```diff
@@ -47,74 +56,74 @@ instance ToRad Ann.WithPos a => MimeRender RadicleData a where
...
+  contentType _ = "application" HttpMedia.// "radicle-json"
+
+newtype NoRadicleJson = NoRadicleJson
+  { error :: Text
+  } deriving (Generic, A.ToJSON)
+
+instance (Traversable t, A.ToJSON (t A.Value)) => MimeRender RadicleJSON (t Value) where
+  mimeRender _ x = case traverse maybeJson x of
+    Right y -> A.encode y
+    Left e -> A.encode (NoRadicleJson ("Radicle value did not have a JSON representation: " <> e))
```
Comment (jkarni) [inline src/Radicle/Daemon/HttpApi.hs]: Why not something like:
```suggestion
    Left e -> A.object [("error", A.String "Radicle value did not have a JSON representation: " <> e)]
```
And then get rid of `NoRadicleJson`?

**Pull request merged**